File: packages/typescript/src/interpreters/voice.ts
/**
 * Voice responder — surfaces the TwiML a voice webhook must return.
 *
 * Does not send anything; lifts the TwiML string out of the adapter's
 * request-description so the webhook layer can return it.
 */
import * as Effect from "effect/Effect"
import type { JsonObject } from "../core/json.ts"
import type { Sent } from "../core/ports.ts"
import type { Transport } from "./transport.ts"

export class VoiceResponder implements Transport {
  dispatch(sent: Sent): Effect.Effect<Sent, never> {
    if (sent.raw.transport !== "twiml") {
      return Effect.succeed(sent)
    }
    const twiml = typeof sent.raw.twiml === "string" ? sent.raw.twiml : ""
    return Effect.succeed({
      ok: true as const,
      message_id: "",
      raw: { native: "twiml", twiml } as JsonObject,
    })
  }
}

File: packages/python/src/caspian/interpreters/voice.py
"""Voice responder — surfaces the TwiML a voice webhook must return.

Voice (Twilio Programmable Voice) is response-based, not push: the developer's
webhook must return a TwiML document in its HTTP response body rather than
making an outbound API call. So this "transport" does not send anything; it
simply lifts the TwiML string out of the adapter's request-description so the
webhook layer can return it. Non-twiml Sents pass through unchanged.
"""

from __future__ import annotations

from caspian.core.ports import Result, Sent


class VoiceResponder:
    """Handles "twiml" Sents by surfacing their TwiML document."""

    def dispatch(self, sent: Sent) -> Result:
        if sent.raw.get("transport") != "twiml":
            return Result.ok(sent)
        return Result.ok(
            Sent(raw={"native": "twiml", "twiml": sent.raw.get("twiml", "")})
        )


__all__ = ["VoiceResponder"]
